File: worker/_shared/illustrations.js
/** EPUB illustration modes — edge port of server/playback/illustrations.py */

export function defaultIllustrationMode(artStyle, imageCount) {
  // "moment" is a manual, opt-in, per-scene feature — never applied
  // automatically. "direct-use" is the only mode that automatically surfaces
  // extracted EPUB plates as cover/character/background art, so when real
  // plates exist, use them rather than silently discarding them into
  // "reference" (server-side i2i hint only, never shown to the user).
  if (imageCount > 0) return "direct-use";
  return "reference";
}

export function normalizeIllustrationMode(mode, artStyle, imageCount) {
  const m = String(mode || "auto").toLowerCase().trim();
  if (["direct", "direct-use", "direct_use", "use"].includes(m)) return "direct-use";
  if (["moment", "flash", "insert"].includes(m)) return "moment";
  if (m === "reference") return "reference";
  return defaultIllustrationMode(artStyle, imageCount);
}

function catalogUrl(catalog, ref) {
  if (ref == null) return null;
  return catalog?.[ref] ?? catalog?.[String(ref)] ?? null;
}

function illustrationCaption(text) {
  const t = String(text || "").trim();
  if (!t) return null;
  return t.length > 72 ? `${t.slice(0, 72).trim()}…` : t;
}

/** First line in reading order spoken by (or, failing that, present with)
 * this character — where a matched plate's "moment" should unlock. Prefers
 * an actual dialogue/narration line credited to the character over a mere
 * presence entry, since that's the more narratively meaningful reveal. */
function firstLineForCharacter(playback, charId) {
  let presentFallback = null;
  for (const scene of playback?.scenes || []) {
    for (const line of scene.lines || []) {
      if (line.character_id === charId) return line;
      if (!presentFallback && (scene.present || []).some((p) => p.character_id === charId)) {
        presentFallback = line;
      }
    }
  }
  return presentFallback;
}

/**
 * Map illustration_ref → book cover / background slots, and character
 * illustration_ref → an unlocked illustration "moment" on that character's
 * first line — NEVER onto character.sprite. A raw EPUB plate (whole page,
 * often multi-character, sometimes a caption-sheet montage) is not a
 * portrait — the portrait a listener sees on stage must stay either the
 * placeholder gradient or actual generated art (reference-conditioned via
 * character.reference_images, see reference-images.js), same as this
 * codebase already treats reference_images as generation input, never
 * display art (see illustration-character-match-consumer.js). The plate
 * itself belongs in the Illustrations gallery / as a story moment instead —
 * see illustrationGallery.js's collectIllustrations, which reads exactly
 * the inserts/line.illustration_url fields set below.
 */
export function applyDirectIllustrations(playback, analysis, illustrationUrls) {
  const counts = { characters: 0, backgrounds: 0, cover: 0 };
  if (!playback || !illustrationUrls || !Object.keys(illustrationUrls).length) {
    return { playback, counts };
  }

  // An explicit cover_illustration_ref (model-matched at extraction time, or
  // user-assigned via the "Cover reference plate" dropdown) always wins —
  // this used to be entirely unread here, so assigning a cover plate saved
  // successfully but never actually changed playback.cover; the loops below
  // only ever set a cover *opportunistically*, as a fallback reusing
  // whichever character/scene plate happened to match first.
  const explicitCoverUrl = catalogUrl(illustrationUrls, analysis?.cover_illustration_ref);
  let coverUrl = explicitCoverUrl || (playback.cover && String(playback.cover).startsWith("/media/")
    ? playback.cover
    : null);

  playback.inserts = playback.inserts || {};
  for (const c of analysis?.characters || []) {
    if (!c?.id) continue;
    const url = catalogUrl(illustrationUrls, c.illustration_ref);
    if (!url || !playback.characters?.[c.id]) continue;
    const line = firstLineForCharacter(playback, c.id);
    if (line) {
      line.illustration_url = url;
      line.illustration_caption = line.illustration_caption || illustrationCaption(line.text);
      line.visual_moment = true;
      playback.inserts[String(line.idx)] = url;
      counts.characters += 1;
    }
    if (!coverUrl) coverUrl = url;
  }

  for (const s of analysis?.scenes || []) {
    if (s.reuse_background_of) continue;
    const url = catalogUrl(illustrationUrls, s.illustration_ref);
    if (!url) continue;
    const sceneOut = (playback.scenes || []).find((sc) => sc.id === s.id);
    if (!sceneOut) continue;
    sceneOut.background = url;
    counts.backgrounds += 1;
    if (!coverUrl) coverUrl = url;
  }

  if (coverUrl && (explicitCoverUrl || counts.characters + counts.backgrounds > 0)) {
    playback.cover = coverUrl;
    counts.cover = 1;
  }

  return { playback, counts };
}

File: legacy/server/playback/illustrations.py
"""EPUB illustration modes: flash moments, sprite-gen reference, legacy direct-use."""
from __future__ import annotations

from ..analyze.schema import BookAnalysis


def default_illustration_mode(art_style: str, image_count: int) -> str:
    """moment = timed flash + per-character gen reference (default when art exists).

    Multi-character inserts are poor permanent sprites but great as brief flashes
    and as reference plates for generating individual sprites.
    """
    if image_count <= 0:
        return "reference"
    if art_style in ("anime", "cartoon"):
        return "moment"
    return "reference"


def normalize_illustration_mode(mode: str | None, art_style: str, image_count: int) -> str:
    m = (mode or "auto").lower().strip()
    if m in ("direct", "direct-use", "direct_use", "use"):
        return "direct-use"
    if m in ("moment", "flash", "insert"):
        return "moment"
    if m == "reference":
        return "reference"
    return default_illustration_mode(art_style, image_count)


def catalog_from_urls(image_urls: dict[int, str]) -> list[str]:
    """Ordered illustration catalog for playback (index → public URL)."""
    if not image_urls:
        return []
    return [image_urls[i] for i in sorted(image_urls)]


def reference_bytes_for_character(
    character_id: str,
    analysis: BookAnalysis,
    all_images: list[bytes] | None,
) -> list[bytes] | None:
    """Pick the best reference plate for sprite generation (single character)."""
    if not all_images:
        return None
    by_id = {c.id: c for c in analysis.characters}
    c = by_id.get(character_id)
    if not c:
        return all_images[:3]
    ref = getattr(c, "illustration_ref", None)
    if ref is not None and 0 <= ref < len(all_images):
        return [all_images[ref]]
    return all_images[:3]


def apply_direct_illustrations(
    book_id: str,
    analysis: BookAnalysis,
    image_urls: dict[int, str],
    *,
    style: str,
    set_media,
) -> dict[str, int]:
    """Legacy: map illustration_ref → permanent media slots (direct-use mode)."""
    counts = {"characters": 0, "backgrounds": 0, "cover": 0}
    if not image_urls:
        return counts

    cover_url = None
    for c in analysis.characters:
        ref = getattr(c, "illustration_ref", None)
        if ref is None or ref not in image_urls:
            continue
        url = image_urls[ref]
        set_media(book_id, "characters", c.id, url, style=style)
        counts["characters"] += 1
        if cover_url is None:
            cover_url = url

    for s in analysis.scenes:
        if s.reuse_background_of:
            continue
        ref = getattr(s, "illustration_ref", None)
        if ref is None or ref not in image_urls:
            continue
        url = image_urls[ref]
        set_media(book_id, "backgrounds", s.id, url, style=style)
        counts["backgrounds"] += 1
        if cover_url is None:
            cover_url = url

    if cover_url and counts["characters"] + counts["backgrounds"] > 0:
        set_media(book_id, "cover", "cover", cover_url, style=style)
        counts["cover"] = 1

    return counts


def resolve_line_illustration(
    line_ref: int | None,
    scene_ref: int | None,
    *,
    is_first_line_in_scene: bool,
    catalog: list[str],
) -> tuple[int | None, str | None]:
    """Line event wins; else scene insert on first line of scene."""
    ref = line_ref
    if ref is None and is_first_line_in_scene:
        ref = scene_ref
    if ref is None or ref < 0 or ref >= len(catalog):
        return None, None
    return ref, catalog[ref]
